plugin/hpsa/hpsa.py

```diff
@@ -909,3 +933,223 @@ def volume_ident_led_clear(self, volume, flags=Client.FLAG_RSVD):
                 "Volume not found")
 
         return None
+
+    @_handle_errors
+    def batteries(self, search_key=None, search_value=None,
+                  flags=Client.FLAG_RSVD):
+        lsm_batteries = []
+        ctrl_all_show = self._sacli_exec(
+            ["ctrl", "all", "show", "config", "detail"])
+
+        for ctrl_name in ctrl_all_show.keys():
+            ctrl_data = ctrl_all_show[ctrl_name]
+            if 'Battery/Capacitor Count' not in ctrl_data or \
+               ctrl_data['Battery/Capacitor Count'] == 0:
+                continue
+
+            sys_id = ctrl_data["Serial Number"]
+
+            battery_type = Battery.TYPE_UNKNOWN
+            battery_status = _hp_battery_status_to_lsm(ctrl_data)
+
+            if ctrl_data["Cache Backup Power Source"] == "Capacitors":
+                battery_type = Battery.TYPE_CAPACITOR
+            elif ctrl_data["Cache Backup Power Source"] == "Batteries":
+                battery_type = Battery.TYPE_CHEMICAL
+            else:
+                raise LsmError(
+                    ErrorNumber.PLUGIN_BUG,
+                    "Got unknown battery type: %s for system %s" %
+                    (ctrl_data["Cache Backup Power Source:"], sys_id))
+
+            for counter in range(0, int(ctrl_data["Battery/Capacitor Count"])):
+
+                lsm_batteries.append(
+                    Battery(
+                        "%s_BAT_%d" % (sys_id, counter),
+                        "Battery %d of %s" % (counter, ctrl_name),
+                        battery_type, battery_status, sys_id, None))
+
+        return lsm_batteries
+
+    @_handle_errors
+    def volume_cache_info(self, volume, flags=Client.FLAG_RSVD):
+        """
+        Depend on command:
+            hpssacli ctrl slot=0 show config detail
+        """
+        flag_battery_ok = False
+        flag_ram_ok = False
+
+        (ctrl_num, array_num, ld_num) = self._cal_of_lsm_vol(volume)
+        ctrl_data = self._sacli_exec(
+            ["ctrl", "slot=%s" % ctrl_num, "show", "config", "detail"]
+            ).values()[0]
+
+        lsm_bats = self.batteries()
+        for lsm_bat in lsm_bats:
+            if lsm_bat.status == Battery.STATUS_OK:
+                flag_battery_ok = True
```

Comment:
@joehandzik Could you answer these questions:
 * Can hpsa support multiple batteries or mix type(capacitor and battery) batteries?
 * Will hpsa enable write cache if one of batteries is in good mode while others don't?

Thank you.